### PR TITLE
New Sections for Custom Certificates

### PR DESCRIPTION
**Pull request by mburke5678**
Based on 
http://v1.uncontained.io/playbooks/installation/custom_certificates.html#option-2-configuring-openshift-to-use-component-specific-custom-certificates

https://github.com/openshift/openshift-docs/issues/7495

### CODE REVIEW

**Comment by vikram-redhat** [comment]
@mburke5678 - is this still coming along? Also note Scott's comment here: https://github.com/openshift/openshift-docs/issues/7495#issuecomment-364319969

**Comment by mburke5678** [comment]
**# REMOVE assetConfig.publicURL for 3.9 PER:**
https://github.com/openshift/openshift-docs/pull/8528

**Review by sdodson — CHANGES_REQUESTED** [review]
i think this looks good after my suggested edits

**Comment by ikke-t** [comment]
To me it looks like it's still talking about cafile in the texts. It could be that I'm reading the diffs wrong, I don't get where I can read the full new version. But I noticed cafile is not utilised when you set new certs, it looks like the latest version here still has cafile in *_named_* certs. The documentation changes actually remove the guidance that cafile is optional.

See my original comment in the issue: "you can leave out cafile parameter here, as it actually does nothing in this case."

**Review by ikke-t — COMMENTED** [review]
Named certificates being optional is removed from comments, and back in the ansible configs, why? It's unnecessary, right?

**Comment by mburke5678** [comment]
@ikke-t Here is a link to a preview page, if it helps. I have made no changes  since you and scott looked at the diffs. 
http://file.rdu.redhat.com/mburke/pr-7495/install_config/certificate_customization.html

File: install_config/certificate_customization.adoc [hunk 1]
```diff
@@ -19,82 +19,248 @@ This can be done during an
...
-openshift_master_named_certificates=[{"certfile": "/path/on/host/to/custom1.crt", "keyfile": "/path/on/host/to/custom1.key", "cafile": "/path/on/host/to/custom-ca1.crt"}]
-#
-# Detected names may be overridden by specifying the "names" key
-#openshift_master_named_certificates=[{"certfile": "/path/on/host/to/custom1.crt", "keyfile": "/path/on/host/to/custom1.key", "names": ["public-master-host.com"], "cafile": "/path/on/host/to/custom-ca1.crt"}]
+openshift_master_named_certificates=[{"names": ["master.148.251.233.173.nip.io"], "certfile": "/home/cloud-user/master-bundle.cert.pem", "keyfile": "/home/cloud-user/master.148.251.233.173.nip.io.key.pem" ]
+----
+
+Example parameters for a wildcard API certificate:
+----
+openshift_master_overwrite_named_certificates=true
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply install_config/certificate_customization.adoc]: @sdodson 
Is `openshift_master_overwrite_named_certificates=true` not required for wildcard certs?

**Comment by mburke5678** [comment]
@ikke-t 
Which use case does not need the ca cert specified? Is the use case for external users (under _Using a Master Host Certificate_ in my draft)?

Also, in your issue, does this string need a closing _}_ after the _.pem"_
openshift_master_named_certificates=[{"names": ["master.148.251.233.173.nip.io"], "certfile": "/home/cloud-user/master-bundle.cert.pem", "keyfile": "/home/cloud-user/master.148.251.233.173.nip.io.key.pem" ]

**Comment by mburke5678** [comment]
@gpei Can you take a look at this PR as you reviewed the original issue?

**Comment by ikke-t** [comment]
@mburke5678 good spot, there was closing missing. About cafile, I believe it was not used by ansibles when running the cert changes, thus is just confusing in the instructions.

File: install_config/certificate_customization.adoc [hunk 2]
```diff
@@ -19,82 +19,277 @@ This can be done during an
...
-#openshift_master_named_certificates=[{"certfile": "/path/on/host/to/custom1.crt", "keyfile": "/path/on/host/to/custom1.key", "names": ["public-master-host.com"], "cafile": "/path/on/host/to/custom-ca1.crt"}]
+openshift_master_named_certificates=[{"names": ["master.148.251.233.173.nip.io"], "certfile": "/home/cloud-user/master-bundle.cert.pem", "keyfile": "/home/cloud-user/master.148.251.233.173.nip.io.key.pem" ]
+----
+
+Example parameters for a wildcard API certificate:
+----
+openshift_hosted_router_certificate={"certfile": "/home/cloud-user/star-apps.148.251.233.173.nip.io.cert.pem", "keyfile": "/home/cloud-user/star-apps.148.251.233.173.nip.io.key.pem", "cafile": "/home/cloud-user/ca-chain.cert.pem"}
 ----
 
 [[configuring-custom-certificates]]
```
Comment (gpei) [inline install_config/certificate_customization.adoc]: Not sure this PR is targeted to which OCP version, for ocp-3.9, just as I comment in previous doc PR before, "assetConfig" field is removed from master config yaml

File: install_config/certificate_customization.adoc [hunk 3]
```diff
@@ -19,82 +19,277 @@ This can be done during an
...
+
+* *keyfile* contains the {product-title} wildcard key
+
+* *cafile* contains the root CA for this key and certificate. If an intermediate CA is in use it should contain both the intermediate and root CA.
+
+If these files are new to your {product-title} cluster, run an Ansible playbook to add these files to the {product-title} configuration files. 
+The playbook adds the certificate files to the *_/etc/origin/master/named_certificates/_* directory.
+
+----
+ansible-playbook /usr/share/ansible/openshift-ansible/playbooks/byo/openshift-cluster/config.yml
```
Comment (gpei) [inline install_config/certificate_customization.adoc]: Not sure this is a supported usage for ocp, we never use installation playbook to add the certificate files

File: install_config/certificate_customization.adoc [hunk 4]
```diff
@@ -19,82 +19,277 @@ This can be done during an
...
+ansible-playbook /usr/share/ansible/openshift-ansible/playbooks/byo/openshift-cluster/redeploy-master-certificates.yml
+----
+
+To retrofit custom router certificates into an existing cluster: 
+
+. Edit the Ansible inventory file *_/etc/ansible/hosts_* to set the `openshift_master_overwrite_named_certificates=true` and specify the path to the certificate using `openshift_hosted_router_certificate`
++
+[source,yaml]
+----
+openshift_master_overwrite_named_certificates=true
```
Comment (gpei) [inline install_config/certificate_customization.adoc]: I'm not familiar with openshift_hosted_router_certificate option, but from my understanding, "openshift_master_overwrite_named_certificates=true" is not used for updating custom router certificates.

File: install_config/certificate_customization.adoc [hunk 5]
```diff
@@ -19,82 +19,277 @@ This can be done during an
...
+openshift_hosted_router_certificate=[{"certfile": "/path/on/host/to/app-crt-file", "keyfile": "/path/on/host/to/app-key-file", "cafile": "/path/on/host/to/app-ca-file"}] <1>
+----
++
+<1> Path to a xref:configuring-custom-certificates-wildcard[wildcard API certificate].
+
+. Run the following playbook:
++
+----
+ansible-playbook /usr/share/ansible/openshift-ansible/playbooks/byo/openshift-cluster/redeploy-router-certificates.yml
```
Comment (gpei) [inline install_config/certificate_customization.adoc]: And for all the playbooks, if it's for ocp-3.9, the path of playbooks are all changed.

**Review by gpei — COMMENTED** [review]
Hi, I found the new added sections are mainly about custom router and registry cert, which is not related to openshift_master_named_certificates, I hardly ever tried with it, so maybe you could find someone else who is familiar with that for a better review, thanks.

**Comment by mburke5678** [comment]
@stuartchuan Is this something you can QE review? If not, can you suggest an appropriate person?

**Comment by stuartchuan** [comment]
@mburke5678 Please see the comments from @gpei .
And for the router and registry cert, please ask @bmeng and @wzheng1 to review.

File: install_config/certificate_customization.adoc (hunk 3 above)
Comment (mburke5678) [reply install_config/certificate_customization.adoc]: This came from @ikke-t issue https://github.com/openshift/openshift-docs/issues/7495. Please advise what is the correct playbook. redeploy-certificates.yml?
"Now as there are new files introduced to OpenShift you need to run ansible configure to make it configure the OpenShift configuration files. The cert files will end up into master's /etc/origin/master/named_certificates/ directory.
ansible-playbook /usr/share/ansible/openshift-ansible/playbooks/byo/openshift-cluster/config.yml"

File: install_config/certificate_customization.adoc (hunk 4 above)
Comment (mburke5678) [reply install_config/certificate_customization.adoc]: @gpei 
These options also came from @ikke-t issue https://github.com/openshift/openshift-docs/issues/7495 and from @sdodson comment https://github.com/openshift/openshift-docs/issues/7495#issuecomment-364319969
Again, please correct me if I have the incorrect options.

**Comment by mburke5678** [comment]
@bmeng and @wzheng1 PTAL.

File: install_config/certificate_customization.adoc [hunk 6]
```diff
@@ -19,82 +19,277 @@ This can be done during an
...
+
+The {product-title} image registry is an internal service that facilitates builds and deployments. Most of the communication with the registry is handled by internal components in {product-title}. As such, there should be no need to replace the certificate used by the registry service itself. 
+
+However, by default, the registry uses routes to allow external systems and users to do pulls and pushes of images. You can use a _re-encrypt route_ with a custom certificate that will be presented to external users instead of using the internal, self-signed certificate. 
+
+To configure this, add the following lines *_/etc/ansible/hosts_* file in the *[OSEv3:vars]*
+section. pointing to the certificates you would like used for the registry route.
+
+----
+openshift_hosted_registry_routehost=registry.apps.c1-ocp.myorg.com <1>
```
Comment (dongboyan77) [inline install_config/certificate_customization.adoc]: Better to add an env var `openshift_hosted_registry_routetermination` here

Comment (mburke5678) [reply install_config/certificate_customization.adoc]: @ikke-t @sdodson Any thoughts on using `openshift_hosted_registry_routetermination` in place of `openshift_hosted_registry_routehost`? If so, `passthrough`or  'reencrypt`?

Comment (ikke-t) [reply install_config/certificate_customization.adoc]: I have no experience on it, I can't help

**Comment by mburke5678** [comment]
@gpei I removed the `assetConfig.publicURL` and updated the playbook paths to 3.9. Can you PTAL.

I will replace these in 3.7 through a follow-up.

**Comment by gpei** [comment]
@mburke5678 Sorry, would you mind pointing me the related change? I checked in this PR, didn't found commit about it.

**Comment by mburke5678** [comment]
@gpei Sorry to be vague. I was asking about [dongboyan77's comment](https://github.com/openshift/openshift-docs/pull/7607#discussion_r180714902) above about replacing `openshift_hosted_registry_routehost=registry.apps.c1-ocp.myorg.com` at Line 236 with 
`openshift_hosted_registry_routetermination`.

**Comment by dongboyan77** [comment]
@mburke5678 I mean keep `openshift_hosted_registry_routehost` , and add another var `openshift_hosted_registry_routetermination` . If you mentioned `reencrypt` route above, then I suggest set it as the value

**Comment by mburke5678** [comment]
@dongboyan77 @gpei  I added `openshift_hosted_registry_routetermination=reencrypt` PTAL

File: install_config/certificate_customization.adoc (hunk 2 above)
Comment (mburke5678) [reply install_config/certificate_customization.adoc]: @gpei I removed `assetConfig` and replaced with `servingInfo`. PTAL particularly the sample code under _.Custom Certificates Configuration_

**Comment by dongboyan77** [comment]
L233-263  LGTM

**Comment by mburke5678** [comment]
@openshift/team-documentation PTAL. A couple of outstanding questions for QE that shouldn't affect the review greatly.

**Comment by mburke5678** [comment]
Note to self: Replace `assetConfig.publicURL` and playbook paths to 3.7 
The `namedCertificates` section should only be configured for the host name
associated with the `masterPublicURL`, **`assetConfig.publicURL`,** and
`oauthConfig.assetPublicURL` settings.

File: install_config/certificate_customization.adoc [hunk 7]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+as if the certificates were new.
+
+[[configuring-custom-certificates-registry]]
+== Configuring a Custom Certificate for the Image Registry
+
+The {product-title} image registry is an internal service that facilitates builds and deployments. Most of the communication with the registry is handled by internal components in {product-title}. As such, there should be no need to replace the certificate used by the registry service itself. 
+
+However, by default, the registry uses routes to allow external systems and users to do pulls and pushes of images. You can use a _re-encrypt route_ with a custom certificate that will be presented to external users instead of using the internal, self-signed certificate. 
+
+To configure this, add the xref:../install_config/install/advanced_install.adoc#advanced-install-configuring-docker-route[following lines]
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: s/this/this route
s/following lines/following lines of code (or text)

File: install_config/certificate_customization.adoc [hunk 8]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+
+[[configuring-custom-certificates-registry]]
+== Configuring a Custom Certificate for the Image Registry
+
+The {product-title} image registry is an internal service that facilitates builds and deployments. Most of the communication with the registry is handled by internal components in {product-title}. As such, there should be no need to replace the certificate used by the registry service itself. 
+
+However, by default, the registry uses routes to allow external systems and users to do pulls and pushes of images. You can use a _re-encrypt route_ with a custom certificate that will be presented to external users instead of using the internal, self-signed certificate. 
+
+To configure this, add the xref:../install_config/install/advanced_install.adoc#advanced-install-configuring-docker-route[following lines] 
+to the Ansible inventory file, by default *_/etc/ansible/hosts_* file, in the `[OSEv3:vars]`
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: "to the `[OSEv3:vars]` section of the Ansible inventory file, by default *_/etc/ansible/hosts_* file. Specify the certificates to use with the registry route."

File: install_config/certificate_customization.adoc [hunk 9]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+section. pointing to the certificates you would like used for the registry route.
+
+----
+openshift_hosted_registry_routehost=registry.apps.c1-ocp.myorg.com <1>
+openshift_hosted_registry_routecertificates={"certfile": "/path/to/registry.apps.c1-ocp.myorg.com.crt", "keyfile": "/path/to/registry.apps.c1-ocp.myorg.com.key", "cafile": "/path/to/registry.apps.c1-ocp.myorg.com-ca.crt"} <2>
+openshift_hosted_registry_routetermination=reencrypt <3>
+----
+
+//https://github.com/openshift/openshift-docs/issues/5765
+<1> The host name of the registry
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: s/registry/registry.

File: install_config/certificate_customization.adoc [hunk 10]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+<2> The locations of the *cacert*, *root*, and *cafile* files.
++
+* *certfile* contains the {product-title} custom router certificate.
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: see previous. :)

File: install_config/certificate_customization.adoc [hunk 11]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+
+* *cafile* contains the root CA for this key and certificate. If an intermediate CA is in use it should contain both the intermediate and root CA.
+<3> Specify where encryption is performed:
++
+* Set to `reencrypt` with a _re-encrypt route_ to terminate encryption at the edge router and re-encrypt it with a new certificate supplied by the destination.
++
+* Set to `passthrough` to terminate encryption at the destination. The destination is responsible for decrypting traffic.
+
+[[configuring-custom-certificates-retrofit]]
+== Retrofit Custom Certificates into a Cluster
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: "retrofit" feels like jargon here. Are we adding or replacing instead?

File: install_config/certificate_customization.adoc [hunk 12]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+
+You can retrofit custom master and custom router ceritificates into an existing {product-title} cluster by editing the
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: certificates

File: install_config/certificate_customization.adoc [hunk 13]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
++
+* Set to `reencrypt` with a _re-encrypt route_ to terminate encryption at the edge router and re-encrypt it with a new certificate supplied by the destination.
++
+* Set to `passthrough` to terminate encryption at the destination. The destination is responsible for decrypting traffic.
+
+[[configuring-custom-certificates-retrofit]]
+== Retrofit Custom Certificates into a Cluster
+
+You can retrofit custom master and custom router ceritificates into an existing {product-title} cluster by editing the 
+the Ansible inventory file, by default *_/etc/ansible/hosts_* and running a playbook.
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: Add a comma after the path.

File: install_config/certificate_customization.adoc [hunk 14]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+
+[[configuring-custom-certificates-retrofit-master]]
+=== Retrofit Custom Master Certificates into a Cluster
+
+To retrofit custom certificates: 
+
+. Edit the Ansible inventory file to set the `openshift_master_overwrite_named_certificates=true` and specify the path to the certificate using `openshift_master_named_certificates`.
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: s/ using `openshift_master_named_certificates`./ as the value of the `openshift_master_named_certificates` parameter.

File: install_config/certificate_customization.adoc [hunk 15]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+----
+ansible-playbook /usr/share/ansible/openshift-ansible/playbooks/redeploy-certificates.yml
+----
+
+[[configuring-custom-certificates-retrofit-router]]
+=== Retrofit Custom Router Certificates into a Cluster
+
+To retrofit custom router certificates: 
+
+. Edit the Ansible inventory file to set the `openshift_master_overwrite_named_certificates=true` and specify the path to the certificate using `openshift_hosted_router_certificate`.
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: s/ using `openshift_hosted_router_certificate`./ as the value of the `openshift_hosted_router_certificate` parameter.

File: install_config/certificate_customization.adoc [hunk 16]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+. Run the following playbook:
++
+----
+ansible-playbook /usr/share/ansible/openshift-ansible/playbooks/redeploy-router-certificates.yml
+----
+
+[[ansible-configuring-custom-certificates-other]]
+== Using Custom Certificates with Other Components
+
+For other components like Logging & Metrics, the approach for fronting services with custom certificates,
```
Comment (kalexand-rh) [inline install_config/certificate_customization.adoc]: I don't understand this line. Do you mean "To provide custom certificates for other internal services, like logging and metrics,"

File: install_config/certificate_customization.adoc [hunk 17]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+The Subject Alternative Names (SAN) should contain the following.
+
+* Two service nostnames:
++
+----
+docker-registry.default.svc.cluster.local
+docker-registry.default.svc            
+----
+
+* Service IP address. Use the following command to get the Docker registry service IP address:
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply install_config/certificate_customization.adoc]: I added the example values to use to match a similar procedure here:
https://docs.openshift.org/latest/install_config/registry/deploy_registry_existing_clusters.html#securing-the-registry-console

File: install_config/certificate_customization.adoc [hunk 18]
```diff
@@ -29,72 +29,294 @@ custom certificates can be configured using the
...
+      - certFile: wildcard.example.com.crt <2>
+        keyFile: wildcard.example.com.key <2>
+        names:
+          - "openshift.example.com"
+  metricsPublicURL: "https://metrics.os.example.com/hawkular/metrics"
+
 ----
 
-Relative paths are resolved relative to the master configuration file. Restart
+<1> Path to cerificate and key files for the CLI and other API calls.
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply install_config/certificate_customization.adoc]: Sorry for the typos. Must have neglected to spell check. Thanks!

File: install_config/certificate_customization.adoc [hunk 19]
```diff
@@ -29,72 +29,300 @@ custom certificates can be configured using the
...
+
+Where the parameter values are: 
+
+* *certfile* contains the {product-title} router certificate.
+
+* *keyfile* contains the {product-title} wildcard key
+
+* *cafile* contains the root CA for this key and certificate. If an intermediate CA is in use it should contain both the intermediate and root CA.
+
+If these certificate files are new to your {product-title} cluster, run an Ansible playbook to add these files to the {product-title} configuration files.
```
[the remark this answers is not in the document]
Comment (mburke5678) [reply install_config/certificate_customization.adoc]: Why change to deployment playbook? I did a search in Atom and see that phrase used only in the 3.6 release notes.

[the remark this answers is not in the document]
Comment (kalexand-rh) [reply install_config/certificate_customization.adoc]: @mburke5678, the pedant in me wants to point out *which* specific playbook you need to run. :)

**Comment by mburke5678** [comment]
Merging 3.7+ based on @ikke-t [comment](https://github.com/openshift/openshift-docs/issues/7495#issuecomment-363485468)

**Comment by mburke5678** [comment]
3.7 https://github.com/openshift/openshift-docs/pull/9029

**Comment by mburke5678** [comment]
3.7 https://github.com/openshift/openshift-docs/pull/9029
3:10 https://github.com/openshift/openshift-docs/pull/9031

**Comment by mburke5678** [comment]
@vikram-redhat @bfallonf I changed the link to the same file in the Day 2 Guide. I hadn't realized the file moved and was not aware of the redirect that made the link work in preview. 
https://github.com/openshift/openshift-docs/pull/9098

**Pull request merged**